File: 2021/Day17/Binary Tree Level Order Traversal/Binary Tree Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
#generate queue and append root to queue
#while there are nodes in the queue,
#iterate through the number of nodes in the queue
#(this is the number of nodes in the next level)
#pop queue(0), add queue(0).val to current ans list and append queue(0)'s children to queue
#append ans to final_list and reset ans = []
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return root
        queue = []
        return_list = []
        queue.append(root)
        while len(queue) > 0:
            ans = []
            l = len(queue)
            for l in range(l):
                node = queue.pop(0)
                ans.append(node.val)
                if node.left != None:
                    queue.append(node.left)
                if node.right != None:
                    queue.append(node.right)
            return_list.append(ans)
        return return_list

File: 2021/Day17/Binary Tree Level Order Traversal/Binary Tree Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    vector<vector<int>> levelOrder(TreeNode *root)
    {

        // Early return condition.
        if (!root)
            return {};

        vector<vector<int>> ourRes; // Carries our result
        vector<int> singleIter;     // Carries a single iteration
                                    // Store the node pointers for the most flexibility.
        queue<TreeNode *> ourNodes;

        // Seed with initial root value.
        ourNodes.push(root);

        while (!ourNodes.empty())
        {
            // Get the size of the current level
            int size = ourNodes.size();

            // Work on the current level.
            while (size--)
            {
                // Push all nodes for a given level into the singleIter vector
                TreeNode *topElement = ourNodes.front();
                singleIter.push_back(topElement->val);
                ourNodes.pop();
                // Prime their children for the next iteration (put all their children in the queue, if they exist).
                if (topElement->left)
                    ourNodes.push(topElement->left);
                if (topElement->right)
                    ourNodes.push(topElement->right);
            }
            // Store all the nodes of one level in a single grouping and push them into our result
            ourRes.push_back(singleIter);
            // Clear the group for the next iteration.
            singleIter.clear();
        }

        return ourRes;
    }
};